## loc, iloc, ix гэх мэт dataframe-с row, column сонгох method-ууд хэрхэн ажилладаг вэ?

In [1]:
import pandas as pd

In [2]:
fbdata = pd.read_csv('/Users/enod/Development/fb-scrape/facebook-page-post-scraper/156915824368931_facebook_statuses.csv')

loc ерөнхийдөө row, column-г сонгохдоо python лист шиг index-р нь зааж сонгож болно

In [22]:
fbdata.loc[:2 , :]

status_id  \
0  156915824368931_1125851687475335   
1  156915824368931_1125254910868346   
2  156915824368931_1125189747541529   

                                      status_message            link_name  \
0                    Өнөөдөр орой хүлээн авч үзээрэй      Timeline Photos   
1  "Бригад" реалити шоуны оролцогчид Сэлэнгэ айма...                  NaN   
2                                 Бригад реалити шоу  Шууд эфир /туршилт/   

  status_type                                        status_link  \
0       photo  https://www.facebook.com/mongoltv/photos/a.349...   
1       video  https://www.facebook.com/mongoltv/videos/11252...   
2       video  https://www.facebook.com/mongoltv/videos/11251...   

      status_published  num_reactions  num_comments  num_shares  num_likes  \
0  2016-08-05 04:16:11             44             6           1         44   
1  2016-08-04 07:39:04             34             0          12         32   
2  2016-08-04 04:39:58            293            52          39        276   

   num_loves  num_wows  num_hahas  num_sads  num_angrys  
0          0         0          0         0           0  
1          2         0          0         0           0  
2          9         4          1         1           2

Эсвэл index-р нь бус column-аа тодорхой зааж өгч болно. Энэ column-аас тэр column хүртлэхийг сонгох гэх мэт

In [24]:
fbdata.loc[:2 , 'status_message':'status_type']

status_message            link_name  \
0                    Өнөөдөр орой хүлээн авч үзээрэй      Timeline Photos   
1  "Бригад" реалити шоуны оролцогчид Сэлэнгэ айма...                  NaN   
2                                 Бригад реалити шоу  Шууд эфир /туршилт/   

  status_type  
0       photo  
1       video  
2       video

In [31]:
fbdata.loc[fbdata.status_type=="video", 'status_message':'status_type'].head()

status_message            link_name  \
1  "Бригад" реалити шоуны оролцогчид Сэлэнгэ айма...                  NaN   
2                                 Бригад реалити шоу  Шууд эфир /туршилт/   
5  "Бригад" реалити шоуны шинэ дугаар 08-р сарын ...                  NaN   
6         "Авьяаслаг Монголчууд 2016" тун удахгүй...                  NaN   
7  "Бригад" реалити шоуны шинэ дугаар 07-р сарын ...                  NaN   

  status_type  
1       video  
2       video  
5       video  
6       video  
7       video

##### iloc loc-тэй адилхан ч integer-р column, row-оо сонгож болно

###### доорхи жишээнд эхний 3н row-г 0,1,2,4 дэхь column-тай сонгож байна

In [41]:
fbdata.iloc[:3,[0,1,2,4]]

status_id  \
0  156915824368931_1125851687475335   
1  156915824368931_1125254910868346   
2  156915824368931_1125189747541529   

                                      status_message            link_name  \
0                    Өнөөдөр орой хүлээн авч үзээрэй      Timeline Photos   
1  "Бригад" реалити шоуны оролцогчид Сэлэнгэ айма...                  NaN   
2                                 Бригад реалити шоу  Шууд эфир /туршилт/   

                                         status_link  
0  https://www.facebook.com/mongoltv/photos/a.349...  
1  https://www.facebook.com/mongoltv/videos/11252...  
2  https://www.facebook.com/mongoltv/videos/11251...

эсвэл range()-р эхний 3н row-г 0-4 хүртэл column-тай сонгох

In [45]:
fbdata.iloc[:3,0:4]

status_id  \
0  156915824368931_1125851687475335   
1  156915824368931_1125254910868346   
2  156915824368931_1125189747541529   

                                      status_message            link_name  \
0                    Өнөөдөр орой хүлээн авч үзээрэй      Timeline Photos   
1  "Бригад" реалити шоуны оролцогчид Сэлэнгэ айма...                  NaN   
2                                 Бригад реалити шоу  Шууд эфир /туршилт/   

  status_type  
0       photo  
1       video  
2       video

###### ix нь iloc,loc 2н хольц буюу column row сонгохдоо string, integer 2уулангаар нь зааж сонгож болдог

In [47]:
fbdata.ix[1,1]

'"Бригад" реалити шоуны оролцогчид Сэлэнгэ аймагт. /Бүтэн үзэхээр/'

In [48]:
fbdata.ix[1,'status_message']

'"Бригад" реалити шоуны оролцогчид Сэлэнгэ аймагт. /Бүтэн үзэхээр/'

In [49]:
fbdata.ix[1,'status_message':'status_type']

status_message    "Бригад" реалити шоуны оролцогчид Сэлэнгэ айма...
link_name                                                       NaN
status_type                                                   video
Name: 1, dtype: object

### Хэзээ inplace хэрэглэх вэ?

In [52]:
fbdata.shape

(3212, 15)

In [53]:
fbdata.head()

status_id  \
0  156915824368931_1125851687475335   
1  156915824368931_1125254910868346   
2  156915824368931_1125189747541529   
3  156915824368931_1125116847548819   
4  156915824368931_1123607134366457   

                                      status_message            link_name  \
0                    Өнөөдөр орой хүлээн авч үзээрэй      Timeline Photos   
1  "Бригад" реалити шоуны оролцогчид Сэлэнгэ айма...                  NaN   
2                                 Бригад реалити шоу  Шууд эфир /туршилт/   
3                          Өнөөдөр орой манай ТВ-ээр      Timeline Photos   
4                                          Өнөө орой      Timeline Photos   

  status_type                                        status_link  \
0       photo  https://www.facebook.com/mongoltv/photos/a.349...   
1       video  https://www.facebook.com/mongoltv/videos/11252...   
2       video  https://www.facebook.com/mongoltv/videos/11251...   
3       photo  https://www.facebook.com/mongoltv/photos/a.349...   
4       photo  https://www.facebook.com/mongoltv/photos/a.349...   

      status_published  num_reactions  num_comments  num_shares  num_likes  \
0  2016-08-05 04:16:11             44             6           1         44   
1  2016-08-04 07:39:04             34             0          12         32   
2  2016-08-04 04:39:58            293            52          39        276   
3  2016-08-04 01:57:23             63            18           3         59   
4  2016-08-02 00:30:56             56            14           1         56   

   num_loves  num_wows  num_hahas  num_sads  num_angrys  
0          0         0          0         0           0  
1          2         0          0         0           0  
2          9         4          1         1           2  
3          1         1          1         0           1  
4          0         0          0         0           0

In [55]:
fbdata.drop('status_id', axis=1).head()

status_message            link_name  \
0                    Өнөөдөр орой хүлээн авч үзээрэй      Timeline Photos   
1  "Бригад" реалити шоуны оролцогчид Сэлэнгэ айма...                  NaN   
2                                 Бригад реалити шоу  Шууд эфир /туршилт/   
3                          Өнөөдөр орой манай ТВ-ээр      Timeline Photos   
4                                          Өнөө орой      Timeline Photos   

  status_type                                        status_link  \
0       photo  https://www.facebook.com/mongoltv/photos/a.349...   
1       video  https://www.facebook.com/mongoltv/videos/11252...   
2       video  https://www.facebook.com/mongoltv/videos/11251...   
3       photo  https://www.facebook.com/mongoltv/photos/a.349...   
4       photo  https://www.facebook.com/mongoltv/photos/a.349...   

      status_published  num_reactions  num_comments  num_shares  num_likes  \
0  2016-08-05 04:16:11             44             6           1         44   
1  2016-08-04 07:39:04             34             0          12         32   
2  2016-08-04 04:39:58            293            52          39        276   
3  2016-08-04 01:57:23             63            18           3         59   
4  2016-08-02 00:30:56             56            14           1         56   

   num_loves  num_wows  num_hahas  num_sads  num_angrys  
0          0         0          0         0           0  
1          2         0          0         0           0  
2          9         4          1         1           2  
3          1         1          1         0           1  
4          0         0          0         0           0

Дээрхи дээр хэдий status_id баганыг арилгасан ч яг датад бол өөрчлөлт ороогүй байгаа. Шалгая:

In [56]:
fbdata.head()

status_id  \
0  156915824368931_1125851687475335   
1  156915824368931_1125254910868346   
2  156915824368931_1125189747541529   
3  156915824368931_1125116847548819   
4  156915824368931_1123607134366457   

                                      status_message            link_name  \
0                    Өнөөдөр орой хүлээн авч үзээрэй      Timeline Photos   
1  "Бригад" реалити шоуны оролцогчид Сэлэнгэ айма...                  NaN   
2                                 Бригад реалити шоу  Шууд эфир /туршилт/   
3                          Өнөөдөр орой манай ТВ-ээр      Timeline Photos   
4                                          Өнөө орой      Timeline Photos   

  status_type                                        status_link  \
0       photo  https://www.facebook.com/mongoltv/photos/a.349...   
1       video  https://www.facebook.com/mongoltv/videos/11252...   
2       video  https://www.facebook.com/mongoltv/videos/11251...   
3       photo  https://www.facebook.com/mongoltv/photos/a.349...   
4       photo  https://www.facebook.com/mongoltv/photos/a.349...   

      status_published  num_reactions  num_comments  num_shares  num_likes  \
0  2016-08-05 04:16:11             44             6           1         44   
1  2016-08-04 07:39:04             34             0          12         32   
2  2016-08-04 04:39:58            293            52          39        276   
3  2016-08-04 01:57:23             63            18           3         59   
4  2016-08-02 00:30:56             56            14           1         56   

   num_loves  num_wows  num_hahas  num_sads  num_angrys  
0          0         0          0         0           0  
1          2         0          0         0           0  
2          9         4          1         1           2  
3          1         1          1         0           1  
4          0         0          0         0           0

fbdata.drop('status_id', axis=1, inplace=True).head() командыг өгснөөр жинхэнэ датад өөрчлөлт орох болно. Дата алдах зорилгоор команд гүйлгэсэнгүй.

inplace заавал ашиглахгүйгээр зүгээр assignment мөн хийж болно. Жишээ нь: fbdata = fbdata.drop('status_id', axis=1, inplace=True) гэх мэт. Үр дүн inplace ашигласантай адил

## Pandas memory usage хэрхэн багасгах вэ?

dataframe.info() method ашиглан memory usage хамгийн багадаа хэд байгааг мэдэж болно.

In [57]:
fbdata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3212 entries, 0 to 3211
Data columns (total 15 columns):
status_id           3212 non-null object
status_message      2649 non-null object
link_name           2413 non-null object
status_type         3212 non-null object
status_link         2748 non-null object
status_published    3212 non-null object
num_reactions       3212 non-null int64
num_comments        3212 non-null int64
num_shares          3212 non-null int64
num_likes           3212 non-null int64
num_loves           3212 non-null int64
num_wows            3212 non-null int64
num_hahas           3212 non-null int64
num_sads            3212 non-null int64
num_angrys          3212 non-null int64
dtypes: int64(9), object(6)
memory usage: 376.5+ KB


memory_usage='deep' гэж зааж өгснөөр яг хэдий хэмжээний memory ашиглаад байгааг нарийвчлан харж болно. Энд бол 2.5 MB RAM ашиглаж буй

In [58]:
fbdata.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3212 entries, 0 to 3211
Data columns (total 15 columns):
status_id           3212 non-null object
status_message      2649 non-null object
link_name           2413 non-null object
status_type         3212 non-null object
status_link         2748 non-null object
status_published    3212 non-null object
num_reactions       3212 non-null int64
num_comments        3212 non-null int64
num_shares          3212 non-null int64
num_likes           3212 non-null int64
num_loves           3212 non-null int64
num_wows            3212 non-null int64
num_hahas           3212 non-null int64
num_sads            3212 non-null int64
num_angrys          3212 non-null int64
dtypes: int64(9), object(6)
memory usage: 2.5 MB


Эсвэл шууд dataframe-с memory_usage method дуудаад column бүр хэр их memory зарцуулц байгааг харж болно.

In [59]:
fbdata.memory_usage()

Index                  80
status_id           25696
status_message      25696
link_name           25696
status_type         25696
status_link         25696
status_published    25696
num_reactions       25696
num_comments        25696
num_shares          25696
num_likes           25696
num_loves           25696
num_wows            25696
num_hahas           25696
num_sads            25696
num_angrys          25696
dtype: int64

Илүү нарийвчлалтай харья гэвэл:

In [60]:
fbdata.memory_usage(deep=True)

Index                   80
status_id           283468
status_message      993081
link_name           319015
status_type         198895
status_link         370266
status_published    244112
num_reactions        25696
num_comments         25696
num_shares           25696
num_likes            25696
num_loves            25696
num_wows             25696
num_hahas            25696
num_sads             25696
num_angrys           25696
dtype: int64

Дээрхи series учраас sum() дуудаж болно:

In [61]:
fbdata.memory_usage(deep=True).sum()

2640181

In [62]:
sorted(fbdata.status_type.unique())

['event', 'link', 'photo', 'status', 'video']

#### status_type дээр жишээ нь 3200 гариу row бүрт дээрхи string-үүдийн аль нэг нь байгаа. Тэгвэл эдгээр string-үүдийг int байдлаар хадгалвал memory usage харьцангүй багасна учир нь string-с int хамаагүй storage efficient. 

##### Pandas дээр үүнийг хийдэг зориулалтын өгөгдлийн төрөл байдаг. 

In [63]:
fbdata['status_type'] = fbdata['status_type'].astype('category')

In [64]:
fbdata.dtypes

status_id             object
status_message        object
link_name             object
status_type         category
status_link           object
status_published      object
num_reactions          int64
num_comments           int64
num_shares             int64
num_likes              int64
num_loves              int64
num_wows               int64
num_hahas              int64
num_sads               int64
num_angrys             int64
dtype: object

Дээрхээс харвал status_type category төрөл болж хувирсан байна.

In [65]:
fbdata.status_type.head()

0    photo
1    video
2    video
3    photo
4    photo
Name: status_type, dtype: category
Categories (5, object): [event, link, photo, status, video]

Эндээс илүү тодорхой харагдаж байгаа байх. Categories нь дотороо 5н төрөлтэй лист байна.

Нэгэнт category төрөл болоод хувиргасан бол үүнийх нь argument-уудыг мөн дуудаж болно

In [66]:
fbdata.status_type.cat.codes.head()

0    2
1    4
2    4
3    2
4    2
dtype: int8

Дээрхээс харвал codes нь photo-2 video-4 гэх мэт string-г int-р илэрхийлж байгааг харуулж байна. Ийм байдлаар memory usage хэмнэдэг байх нь. Шалгая:

In [69]:
fbdata.memory_usage(deep=True)

Index                   80
status_id           283468
status_message      993081
link_name           319015
status_type           3522
status_link         370266
status_published    244112
num_reactions        25696
num_comments         25696
num_shares           25696
num_likes            25696
num_loves            25696
num_wows             25696
num_hahas            25696
num_sads             25696
num_angrys           25696
dtype: int64

status_type нь 198895 байсан бол category ашигласнаар 3522 болж багассан байна.

#### Энд гол анхаарах зүйлсийн нэг бол category үргэлж memory usage хэмнээд байдаггүй. Жишээ нь бүх status_id-г category болгосон бол хэрэглэж байгаа memory usage улам ихсэнэ. Тэгэхээр category зөвхөн тодорхой хэдэн бичлэгтэй л талбарт ашиглавал ашигтай байна.

### Хэрхэн string утгаар sort_values хийх вэ?

Жишээ df үүсгэе:


In [96]:
df = pd.DataFrame({'ID':[100,101,102,103], 'чанар':["сайн","маш сайн","сайн","төгс"]})

In [97]:
df

ID     чанар
0  100      сайн
1  101  маш сайн
2  102      сайн
3  103      төгс

In [98]:
df.sort_values('чанар')

ID     чанар
1  101  маш сайн
0  100      сайн
2  102      сайн
3  103      төгс

Энд sort хийхдээ үсгийн дарааллар хийсэн байна. Гэхдээ чанараар нь хэрхэн sort хийх вэ?

In [99]:
df['чанар'] = df['чанар'].astype('category', categories=['сайн','маш сайн','төгс'], ordered=True)

In [100]:
df.чанар

0        сайн
1    маш сайн
2        сайн
3        төгс
Name: чанар, dtype: category
Categories (3, object): [сайн < маш сайн < төгс]

Дээрхээс харахад categories нь сайн < маш сайн < төгс байдлаар ангиллагдсан байна.

In [101]:
df.sort_values('чанар')

ID     чанар
0  100      сайн
2  102      сайн
1  101  маш сайн
3  103      төгс

Ингэж болж байгаан бол boolean operator мөн ашиглаад зөвхөн хүссэн чанараа сонгож болох нь

In [103]:
df.loc[df.чанар > 'сайн', :]

ID     чанар
1  101  маш сайн
3  103      төгс